File: internal/protocol/protocol.go
// Package protocol encodes PCsensor configuration payloads, without HID framing.
package protocol

import (
	"fmt"
	"strconv"
	"strings"
)

var modifiers = map[string]byte{"ctrl": 1, "lctrl": 1, "shift": 2, "lshift": 2, "alt": 4, "lalt": 4, "cmd": 8, "super": 8, "win": 8, "lmeta": 8, "rctrl": 16, "rshift": 32, "ralt": 64, "rcmd": 128, "rmeta": 128}
var keys = map[string]byte{"enter": 40, "escape": 41, "esc": 41, "backspace": 42, "tab": 43, "space": 44, "minus": 45, "equal": 46, "leftbracket": 47, "rightbracket": 48, "backslash": 49, "semicolon": 51, "quote": 52, "grave": 53, "comma": 54, "period": 55, "slash": 56, "capslock": 57, "printscreen": 70, "scrolllock": 71, "pause": 72, "insert": 73, "home": 74, "pageup": 75, "delete": 76, "end": 77, "pagedown": 78, "right": 79, "left": 80, "down": 81, "up": 82, "numlock": 83}

// Key accepts modifiers and one USB keyboard usage, independent of OS layout.
func Key(s string) ([]byte, error) {
	p := []byte{8, 1, 0, 0, 0, 0, 0, 0}
	for _, part := range strings.Split(strings.ToLower(s), "+") {
		if m, ok := modifiers[part]; ok {
			if p[2]&m != 0 {
				return nil, fmt.Errorf("duplicate modifier")
			}
			p[2] |= m
			continue
		}
		var k byte
		if len(part) == 1 && part[0] >= 'a' && part[0] <= 'z' {
			k = part[0] - 'a' + 4
		} else if len(part) == 1 && part[0] >= '1' && part[0] <= '9' {
			k = part[0] - '1' + 30
		} else if part == "0" {
			k = 39
		} else if strings.HasPrefix(part, "f") {
			n, e := strconv.Atoi(part[1:])
			if e == nil && n >= 1 && n <= 24 {
				if n <= 12 {
					k = byte(n + 57)
				} else {
					k = byte(n + 91)
				}
			}
		} else {
			k = keys[part]
		}
		if k == 0 || p[3] != 0 {
			return nil, fmt.Errorf("expected modifiers and one supported key")
		}
		p[3] = k
	}
	return p, nil
}

func Validate(p []byte) error {
	// Length is one byte; the first two bytes are record length and action type.
	if len(p) < 2 || len(p) > 255 || int(p[0]) != len(p) {
		return fmt.Errorf("invalid configuration record length")
	}
	return nil
}

// ValidateWritable accepts only action formats established by protocol sources.
// Unknown macros can be read for diagnosis, but cannot be safely replayed.
func ValidateWritable(p []byte) error {
	if e := Validate(p); e != nil {
		return e
	}
	switch p[1] {
	case 0:
		if len(p) == 8 && strings.Trim(string(p[2:]), "\x00") == "" {
			return nil
		}
	case 1, 0x81:
		if (len(p) == 4 || len(p) == 8) && strings.Trim(string(p[4:]), "\x00") == "" {
			return nil
		}
	case 2, 3:
		if len(p) == 8 && p[4] <= 7 {
			return nil
		}
	case 4:
		if len(p) >= 3 && len(p) <= 40 {
			return nil
		}
	}
	return fmt.Errorf("unsupported configuration action or shape; refusing replay")
}

func Query(slot, count int) ([]byte, error) {
	if slot < 1 || slot > count {
		return nil, fmt.Errorf("slot must be between 1 and %d", count)
	}
	return []byte{1, 0x82, 8, byte(slot), 0, 0, 0, 0}, nil
}

func WriteReports(slot, count int, p []byte) ([][]byte, error) {
	if _, e := Query(slot, count); e != nil {
		return nil, e
	}
	if e := ValidateWritable(p); e != nil {
		return nil, e
	}
	reports := [][]byte{{1, 0x81, byte(len(p)), byte(slot), 0, 0, 0, 0}}
	for offset := 0; offset < len(p); offset += 8 {
		b := make([]byte, 8)
		copy(b, p[offset:])
		reports = append(reports, b)
	}
	return reports, nil
}

// Equivalent allows firmware to omit trailing zeroes in fixed keyboard records.
func Equivalent(a, b []byte) bool {
	if Validate(a) != nil || Validate(b) != nil {
		return false
	}
	if a[1] != b[1] {
		return false
	}
	if a[1] != 1 && a[1] != 0x81 {
		return string(a) == string(b)
	}
	return strings.TrimRight(string(a[1:]), "\x00") == strings.TrimRight(string(b[1:]), "\x00")
}

File: web/protocol.mjs
export const models = Object.freeze([
  Object.freeze({name: '400-MA214BK', pid: 0xb001, slots: 3, reportId: 0}),
  Object.freeze({name: '400-SKB081', pid: 0xc115, slots: 6, reportId: 1}),
]);
const modifiers = {ctrl:1,lctrl:1,shift:2,lshift:2,alt:4,lalt:4,cmd:8,super:8,win:8,lmeta:8,rctrl:16,rshift:32,ralt:64,rcmd:128,rmeta:128};
const keys = {enter:40,escape:41,esc:41,backspace:42,tab:43,space:44,minus:45,equal:46,leftbracket:47,rightbracket:48,backslash:49,semicolon:51,quote:52,grave:53,comma:54,period:55,slash:56,capslock:57,printscreen:70,scrolllock:71,pause:72,insert:73,home:74,pageup:75,delete:76,end:77,pagedown:78,right:79,left:80,down:81,up:82,numlock:83};
const owns = (object, key) => Object.hasOwn(object, key);
export function encodeKey(text) {
  if (typeof text !== 'string') throw Error('Enter a keyboard shortcut.');
  const record = Uint8Array.of(8,1,0,0,0,0,0,0);
  for (const part of text.toLowerCase().split('+')) {
    if (owns(modifiers,part)) {
      if (record[2] & modifiers[part]) throw Error('Duplicate modifier.');
      record[2] |= modifiers[part];
      continue;
    }
    let key = 0;
    if (/^[a-z]$/.test(part)) key = part.charCodeAt(0)-93;
    else if (/^[1-9]$/.test(part)) key = Number(part)+29;
    else if (part==='0') key=39;
    else if (/^f([1-9]|1[0-9]|2[0-4])$/.test(part)) {const n=Number(part.slice(1));key=n+(n<=12?57:91);}
    else if (owns(keys,part)) key=keys[part];
    if (!key || record[3]) throw Error('Use modifiers and one supported key, such as ctrl+shift+a.');
    record[3]=key;
  }
  return record;
}
export function validateRecord(record) {
  if (!(record instanceof Uint8Array) || record.length<2 || record.length>255 || record[0]!==record.length) throw Error('Invalid configuration record length.');
  return record;
}
export function validateWritable(record) {
  validateRecord(record);
  const size=record.length, type=record[1];
  if (type===0 && size===8 && record.slice(2).every(n=>n===0)) return record;
  if ((type===1 || type===0x81) && (size===4 || size===8) && record.slice(4).every(n=>n===0)) return record;
  if ((type===2 || type===3) && size===8 && record[4]<=7) return record;
  if (type===4 && size>=3 && size<=40) return record;
  throw Error('Unsupported configuration action or shape; this record cannot be safely restored.');
}
const hex = record => Array.from(record,n=>n.toString(16).padStart(2,'0')).join('');
export function equivalent(a,b) {
  try {validateRecord(a);validateRecord(b);} catch {return false;}
  if (a[1]!==b[1]) return false;
  if (a[1]!==1 && a[1]!==0x81) return hex(a)===hex(b);
  let endA=a.length, endB=b.length;
  while (endA>1 && a[endA-1]===0) endA--;
  while (endB>1 && b[endB-1]===0) endB--;
  return hex(a.slice(1,endA))===hex(b.slice(1,endB));
}
function knownModel(model) {
  const name=typeof model==='string'?model:model?.name;
  const found=models.find(item=>item.name===name);
  if (!found) throw Error('Unsupported model.');
  return found;
}
export function snapshot(model,records) {
  model=knownModel(model);
  if (!Array.isArray(records) || records.length!==model.slots) throw Error('Backup has the wrong number of slots.');
  return {version:1,model:model.name,records:records.map(record=>hex(validateRecord(record)))};
}
export function parseSnapshot(text,expectedModel) {
  // CLI ceiling: six maximum 255-byte records, formatted envelope, final newline.
  const maxSize=JSON.stringify({version:1,model:'400-MA214BK',records:Array(6).fill('ff'.repeat(255))},null,2).length+1;
  if (typeof text!=='string' || text.length>maxSize || new TextEncoder().encode(text).length>maxSize) throw Error('Backup is too large.');
  let value;
  try {value=JSON.parse(text);} catch {throw Error('Backup must contain valid JSON.');}
  const model=knownModel(expectedModel);
  if (!value || typeof value!=='object' || Array.isArray(value) || Object.keys(value).some(key=>!['version','model','records'].includes(key)) || value.version!==1 || value.model!==model.name || !Array.isArray(value.records) || value.records.length!==model.slots) throw Error('Backup version, model, or slot count does not match.');
  return value.records.map(record=>{
    if (typeof record!=='string' || !/^(?:[0-9a-fA-F]{2}){2,255}$/.test(record)) throw Error('Invalid hexadecimal configuration record.');
    return validateWritable(Uint8Array.from(record.match(/../g),pair=>parseInt(pair,16)));
  });
}
export function describeRecord(record) {
  validateRecord(record);
  try {validateWritable(record);} catch {return `hex:${hex(record)}`;}
  if (record[1]===0) return 'disabled';
  if (record[1]!==1 && record[1]!==0x81) return `hex:${hex(record)}`;
  let key;
  const usage=record[3];
  if (usage>=4 && usage<=29) key=String.fromCharCode(usage+93);
  else if (usage>=30 && usage<=38) key=String(usage-29);
  else if (usage===39) key='0';
  else if (usage>=58 && usage<=69) key=`f${usage-57}`;
  else if (usage>=104 && usage<=115) key=`f${usage-91}`;
  else key=Object.keys(keys).find(name=>keys[name]===usage);
  if (!key && usage!==0) return `hex:${hex(record)}`;
  const parts=['ctrl','shift','alt','cmd','rctrl','rshift','ralt','rcmd'].filter((_,i)=>record[2]&(1<<i));
  if (key) parts.push(key);
  return (record[1]===0x81?'release: ':'')+(parts.join('+') || 'no key');
}
